### PR TITLE
Decommission retry/pr

### PR DESCRIPTION
**Pull request by imRishN**
### Description
[Describe what this change achieves]

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java [hunk 1]
```diff
@@ -28,22 +29,34 @@
...
 
-    public DecommissionRequest(DecommissionAttribute decommissionAttribute) {
+    public DecommissionRequest(DecommissionAttribute decommissionAttribute, boolean retryOnClusterManagerChange, TimeValue retryTimeout) {
         this.decommissionAttribute = decommissionAttribute;
+        this.retryOnClusterManagerChange = retryOnClusterManagerChange;
+        this.retryTimeout = retryTimeout;
+    }
+
+    public DecommissionRequest(DecommissionAttribute decommissionAttribute, TimeValue retryTimeout) {
+        this(decommissionAttribute, false, retryTimeout);
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:44]: how are we making sure that user doesn't specify this parameter ? did we evaluate putting this in request context as it is an internal detail of a request and not a user facing one  ?

File: server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java [hunk 2]
```diff
@@ -72,6 +75,57 @@ public class DecommissionController {
...
+            );
+            listener.onFailure(
+                new OpenSearchTimeoutException(
+                    "timed out before retrying [{}] for attribute [{}] after cluster manager change",
+                    DecommissionAction.NAME,
+                    decommissionRequest.getDecommissionAttribute()
+                )
+            );
+            return;
+        }
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:98]: do we need `remainingTimeoutMS` check here only ? Why do we need it ?

File: server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java [hunk 3]
```diff
@@ -210,18 +216,23 @@ public void onResponse(Void unused) {
...
+                                () -> new ParameterizedMessage(
+                                    "failed to retry decommission action for attribute [{}]",
+                                    decommissionRequest.getDecommissionAttribute()
+                                ),
+                                t
+                            );
+                            clearVotingConfigExclusionAndUpdateStatus(false, false); // TODO - need to test this
+                            delegatedListener.onFailure(t);
+                        })
                     );
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:236]: how are responding to user API call now ? Earlier , the retried request on new elected cluster manager  would respond it . Would the same hold true now as well ?

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java (hunk 1 above)
Comment (imRishN) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:44]: Are you talking about the retry flag or the timeout? 
For flag, The REST action doesn't support accepting retry flag with the user and hence any request created at REST layer will always have it set to false. Today, it is only set to true when we trigger retry action from service
For timeout, the default is set and user can set one according to his case

Let me know if this answers your question

File: server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java (hunk 2 above)
Comment (imRishN) [reply server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:98]: This timeout is a check for retry eligibility only. Other actions as part of decommission has their own timeouts. This is the place where we are actually triggering the retry and hence the check. If timed out, request will not be eligible for a retry

File: server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java (hunk 3 above)
Comment (imRishN) [reply server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:236]: Yes, response to the retried request is attached to the listener of original request. The chain will continue if multiple retries gets executed.

File: server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java (hunk 2 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:98]: can we do without this ? In worst case, the retried action will timeout and we will throw a timeout exception . This part of the code is looking out of place from readability POV.

**Review by gbbafna — APPROVED** [review]

**Comment by imRishN** [comment]
Closing this and opened a new PR in upstream repo - https://github.com/opensearch-project/OpenSearch/pull/4684

**Pull request closed without merging**